src/controller/proxy/local.go

```diff
@@ -0,0 +1,184 @@
+//  limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/replication/adapter"
+	"github.com/goharbor/harbor/src/replication/adapter/harbor/base"
+	"github.com/goharbor/harbor/src/replication/adapter/harbor/v2"
+	"github.com/goharbor/harbor/src/replication/model"
+	"io"
+	"time"
+)
+
+// localInterface defines operations related to localHelper repo under proxy mode
+type localInterface interface {
+	// BlobExist check if the blob exist in localHelper repo
+	BlobExist(ctx context.Context, dig string) (bool, error)
+	// PushBlob push blob to localHelper repo
+	PushBlob(localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error
+	// PushManifest push manifest to localHelper repo
+	PushManifest(repo string, tag string, manifest distribution.Manifest) error
+	// PushManifestList push manifest list to localHelper repo
+	PushManifestList(ctx context.Context, repo string, tag string, man distribution.Manifest) error
+	// CheckDependencies check if the manifest's dependency is ready
+	CheckDependencies(ctx context.Context, man distribution.Manifest) []distribution.Descriptor
+	// DeleteManifest cleanup delete tag from localHelper cache
+	DeleteManifest(repo, ref string)
+}
+
+// localHelper defines operations related to localHelper repo under proxy mode
+type localHelper struct {
+	registry adapter.ArtifactRegistry
+}
+
+// newLocalHelper create the localInterface
+func newLocalHelper() (localInterface, error) {
+	l := &localHelper{}
+	if err := l.init(); err != nil {
+		log.Errorf("Failed to init localHelper, error %v", err)
+		return nil, fmt.Errorf("failed to initialize the local helper")
+	}
+	return l, nil
+}
+
+func (l *localHelper) BlobExist(ctx context.Context, dig string) (bool, error) {
+	// not using l.registry.BlobExist(repo, dig)
+	// sometimes the blob exist in storage,
+	// but it does not exist in the Harbor database.
+	// if push manifest to Harbor, it still get failed
+	return blob.Ctl.Exist(ctx, dig)
+}
+
+func (l *localHelper) init() error {
+	if l.registry != nil {
+		return nil
+	}
+	log.Debugf("core url:%s, localHelper core url: %v", config.GetCoreURL(), config.LocalCoreURL())
+	// the traffic is internal only
+	registryURL := config.LocalCoreURL()
+	// TODO: need to verify it works
+	// registryURL := config.GetCoreURL()
+	reg := &model.Registry{
+		URL: registryURL,
+		Credential: &model.Credential{
+			Type:         model.CredentialTypeSecret,
+			AccessSecret: config.ProxyServiceSecret,
+		},
+		TokenServiceURL: registryURL + "/service/token",
+		Insecure:        true,
+	}
+	baseAdapter, err := base.New(reg)
+	if err != nil {
+		return err
+	}
+	adp := v2.New(baseAdapter)
```

Comment:
Use the factory to create the adapter instead

Reply:
change to registry.Client